src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -66,23 +76,81 @@ public async Task<BundleResponse> Handle(BundleRequest bundleRequest, Cancellati
         {
             var bundleResource = bundleRequest.Bundle.ToPoco<Hl7.Fhir.Model.Bundle>();
 
-            if (bundleResource.Type != Hl7.Fhir.Model.Bundle.BundleType.Batch)
+            await FillRequestLists(bundleResource.Entry);
+
+            if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Batch)
+            {
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.BatchResponse,
+                };
+
+                await ExecuteAllRequests(responseBundle);
+                return new BundleResponse(responseBundle.ToResourceElement());
+            }
+            else if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Transaction)
             {
-                throw new MethodNotAllowedException(Microsoft.Health.Fhir.Api.Resources.OnlyCertainBundleTypesSupported);
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.TransactionResponse,
+                };
+
+                return await ExecuteTransactionForAllRequests(responseBundle);
             }
 
-            await FillRequestLists(bundleResource.Entry);
+            throw new MethodNotAllowedException(string.Format(Api.Resources.InvalidBundleTypeError, bundleResource.Type));
+        }
 
-            var responseBundle = new Hl7.Fhir.Model.Bundle
+        private async Task<BundleResponse> ExecuteTransactionForAllRequests(Hl7.Fhir.Model.Bundle responseBundle)
+        {
+            try
             {
-                Type = Hl7.Fhir.Model.Bundle.BundleType.BatchResponse,
-            };
+                using (var transaction = _transactionHandler.BeginTransaction())
+                {
+                    await ExecuteAllRequests(responseBundle);
 
-            await ExecuteAllRequests(responseBundle);
+                    transaction.Complete();
+                }
+            }
+            catch (TransactionAbortedException)
+            {
+                _logger.LogError("Failed to commit a transaction. Throwing BadRequest as a default exception.");
+                throw new BadRequestException(Api.Resources.GeneralTransactionAbortedError);
+            }
 
             return new BundleResponse(responseBundle.ToResourceElement());
         }
 
+        private static void ThrowTransactionException(HttpContext httpContext, Resource responseOperationOutcome)
+        {
+            var operationOutcomeIssues = GetOperationOutcomeIssues(((OperationOutcome)responseOperationOutcome).Issue);
```

Comment:
Can we change the incoming parameter to the type `OperationOutcome` since this is already checked/expected in the code that calls this method?